### DISCUSSION TITLE
Why use dynamic API routes?

### DISCUSSION THREAD

**Discussion by kai-franke** [neuefische/questions#314] [Web] [answered]
Hello!

For the HTTP methods GET and POST I use a simple static API route `/api/tutorials/index.js`.
Now I create a delete functionality and want to remove single entries from my database with the HTTP method DELETE. The other students use a dynamic API route `/api/tutorials/[id].js` for this and it works very well.

Still, I was wondering why I need to transport the ID of the single entry through the route? So I tried to include the DELETE method as a third option in my static API and include the ID in the body of the response. To me it looks like the functionality is identical.

Can someone please explain to me what is the advantage of using a dynamic API route?

These are my two solutions:
**Dynamic API:** https://github.com/kai-franke/capstone-project/blob/feature/delete/pages/api/tutorials/%5Bid%5D.js
**Static API:** https://github.com/kai-franke/capstone-project/blob/feature/delete-static-api/pages/api/tutorials/index.js

Thanks a lot!

**Comment by ThomasF85** [comment, marked as answer]
Hi Kai,

it's about good REST API design, so that other developers and people who use your API will understand directly what the API is doing. REST APIs are intended to do operations on resources (your tutorial objects are resources). The path or url is supposed to identify the resource. The http method (GET, POST, DELETE, ...) is supposed to identify what kind of operation is being performed on the resource. A good API design would look like this:
```
GET /api/tutorials => returns all tutorials
POST /api/tutorials => adds a new tutorial

GET /api/tutorials/[id] => returns tutorial with respective id
PUT /api/tutorials/[id] => modifies(overwrites) tutorial with respective id
PATCH /api/tutorials/[id] => modifies(partially) tutorial with respective id
DELETE /api/tutorials/[id] => deletes tutorial with respective id
```

If you allow the DELETE method on `/api/tutorials` everybody would expect the API call to delete all tutorials.

**Reply by kai-franke** [reply]
That makes perfect sense. Thanks for the quick reply. 🙂
